### PR TITLE
citra_qt: Implement Movie Recording & Playback

### PR DESCRIPTION
**Pull request by valentinvanelslande**
Recording/Playback will start after opening a game.
Recording will stop after stopping the game.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3500)

### CODE REVIEW

File: src/core/movie.cpp [hunk 1]
```diff
@@ -129,6 +129,7 @@ void Movie::CheckInputEnd() {
     if (current_byte + sizeof(ControllerState) > recorded_input.size()) {
         LOG_INFO(Movie, "Playback finished");
         play_mode = PlayMode::None;
+        Settings::values.movie_play = "";
```
Comment (lioncash) [inline src/core/movie.cpp]: `movie_play.clear();`

File: src/core/movie.cpp [hunk 2]
```diff
@@ -388,7 +389,10 @@ void Movie::SaveMovie() {
 
     if (!save_record.IsGood()) {
         LOG_ERROR(Movie, "Error saving movie");
+        return;
     }
+
+    Settings::values.movie_record = "";
```
Comment (lioncash) [inline src/core/movie.cpp]: `movie_record.clear();`

File: src/citra_qt/main.cpp [hunk 3]
```diff
@@ -1048,6 +1056,22 @@ void GMainWindow::OnCreateGraphicsSurfaceViewer() {
     graphicsSurfaceViewerWidget->show();
 }
 
+void GMainWindow::OnRecordMovie() {
+    QString path = QFileDialog::getSaveFileName(this, tr("Save movie"));
+    if (path.isEmpty())
+        return;
+    if (Settings::values.movie_play != "")
```
Comment (lioncash) [inline src/citra_qt/main.cpp]: `!Settings::values.movie_play.empty()`

File: src/citra_qt/main.cpp [hunk 4]
```diff
@@ -1048,6 +1056,22 @@ void GMainWindow::OnCreateGraphicsSurfaceViewer() {
...
+        Settings::values.movie_play = "";
```
Comment (lioncash) [inline src/citra_qt/main.cpp]: .clear()

File: src/citra_qt/main.cpp [hunk 5]
```diff
@@ -1048,6 +1056,22 @@ void GMainWindow::OnCreateGraphicsSurfaceViewer() {
...
+    Settings::values.movie_record = path.toStdString();
+}
+
+void GMainWindow::OnPlayMovie() {
+    QString path = QFileDialog::getOpenFileName(this, tr("Play Movie"));
+    if (Settings::values.movie_record != "")
```
Comment (lioncash) [inline src/citra_qt/main.cpp]: `!Settings::values.movie_record.empty()`

File: src/citra_qt/main.cpp [hunk 6]
```diff
@@ -1048,6 +1056,22 @@ void GMainWindow::OnCreateGraphicsSurfaceViewer() {
...
+        Settings::values.movie_record = "";
```
Comment (lioncash) [inline src/citra_qt/main.cpp]: `.clear()`

File: src/citra_qt/main.cpp [hunk 7]
```diff
@@ -1048,6 +1056,22 @@ void GMainWindow::OnCreateGraphicsSurfaceViewer() {
     graphicsSurfaceViewerWidget->show();
 }
 
+void GMainWindow::OnRecordMovie() {
+    QString path = QFileDialog::getSaveFileName(this, tr("Save movie"));
```
Comment (lioncash) [inline src/citra_qt/main.cpp]: The M in `movie` should be uppercased

**Comment by adityaruplaha** [comment]
Does it start automatically? If so, please change it to a option.
([   ] Record movies automatically)
Other than that, LGTM.

**Comment by Simonx22** [comment]
@adityaruplaha you need to select Movie -> Record first and type in a file name. Then you just start the game and once you're done you stop the emulation. To replay you just go to Movie -> Play and open the file and then the game.

**Comment by adityaruplaha** [comment]
Thanks. I thought it would start automatically whenever a game is opened,
which I don't like. Please edit the PR description.

Adityarup Laha

**Comment by B3n30** [comment]
First of: UI changes require a screenshot. 
Second: Is this really necessary? Movie recording/playback is usually a TAS/Debug feature so it's mainly used with sdl

**Comment by Simonx22** [comment]
Screenshot: 
image

**Comment by BreadFish64** [comment]
@B3n30 I don't see why it shouldn't be in the QT frontend, I was gonna add this myself eventually.

**Comment by adityaruplaha** [comment]
@B3n30

Don't mean to sound rude but who told you that TASers don't use the Qt
frontend? Most users prefer it.

Adityarup Laha

**Comment by jroweboy** [comment]
@adityaruplaha there are no TAS tools for citra right now, so its very pointless to even mention TASers with relation to the movie system. when we have TAS tools, then it becomes relevant. 

I'll give a review of this at some point >.< still kinda bogged down with other things though

**Comment by adityaruplaha** [comment]
Yeah, but users (or TASers) might want to record something (like a graphical glitch) for Debug purposes. That's what I meant actually.

**Comment by adityaruplaha** [comment]
Also, can Movie recording be started/stopped while a game is running? If not so, please add that feature.

**Comment by valentinvanelslande** [comment]
@adityaruplaha no, that requires movie system restart

**Comment by adityaruplaha** [comment]
Well, please implement it. Coz most people don't record full gameplay, only
parts of it.

Adityarup Laha

**Comment by valentinvanelslande** [comment]
someone else will do it, I'm working on HTTP service (using cpp-httplib)

**Comment by adityaruplaha** [comment]
Ok, I'll try once I'm done with #3417 .

Adityarup Laha

**Comment by wwylele** [comment]
What's the status of this?

**Comment by Alkaris** [comment]
How are you supposed to save TAS videos with this? it doesn't even give the file an extensions that Citra can read for this. What file format extension is Citra TAS Movie? I recorded something but it can't be opened to be played back.

**Comment by adityaruplaha** [comment]
@alkaris IIRC TAS videos are saved as .ctm files. They're basically input recordings which simulate user input upon playback. You can't open them in a Media Player, if that's what you were looking for.

**Pull request closed without merging**